RearrangingCars.java

```diff
@@ -0,0 +1,207 @@
+        }
+        
+        //If the initial or expected rearrangement array is not a valid permutation of {0,...,array.length-1}, then throw an exception.
+        else if(!isValidPermutation(initial) || !isValidPermutation(expected)){
+            throw new InvalidRearrangementException("The initial or the expected rearrangement is invalid!");
+        }
+        
+        else{
+            //If the initial and expected rearrangement is empty (no slots), then nothing to do, simply return "".
+            if(initial.length == 0){
+                return str.toString();
+            }
+            
+            //If the initial and expected rearrangement are equal, then no steps required, return "".
+            boolean isEqual = true;
+            for(int i=0; i<initial.length; i++){
+                if(initial[i] != expected[i]){
+                    isEqual = false;
+                }
+            }
+            if(isEqual){
+                return str.toString();
+            }
+            else{
+                //Obtain the required series of steps and print them
+                ArrayList<Step> steps = rearrangingStepsHelper(initial, expected);
+                str.append(steps.toString());
+                        
+                return str.toString();
+            }
+        }
+    }
+    
+    /** rearrangingStepsHelper(current, expected) generates a series of steps required to arrange the cars
+     *  according to the "expected" rearrangement, starting from the "current" rearrangement. 
+     *  At each step, a car is moved from its location to the empty slot.
+     * 
+     * Solution:
+     *  1. Create an inverse mapping of the current rearrangement.
+     *  2. Get the index corresponding to the empty slot, and look for the current location
+     *     of the car that is expected in that slot (using inverse mapping), (let us call it "move_from").
+     *     i) If "move_from" is SAME as "empty_slot" (current and expected both have the same empty slot index),
+     *        then, select a random index (where the car parked differs from expected) and make it "move_from".
+     *        Then, go to step (3).
+     *  3. Make a move from move_from index to empty_slot index, and update the current rearrangement.
+     *  4. Repeat until the current rearrangement becomes same as that of the expected rearrangement.
+     * 
+     * @param current: an array representing current rearrangement of the cars
+     * @param expected: an array representing expected rearrangement of the cars
+     * @return: ArrayList<Steps>: representing the required series of steps
+     */
+    private static  ArrayList<Step> rearrangingStepsHelper(int [] current, int [] expected){
+        
+        ArrayList<Step> steps = new ArrayList<Step>();
+        
+        //Create an inverse-mapping of the "current" rearrangement.
+        int [] inverseMapping_current = getInverseMapping(current);            

```

Comment:
Just as a small comment, you can get around the need of having an inverse mapping array by using a HashMap as Natalija, Nada and I did because for me at least, is a little bit to follow at the moment unless I have a pen and a paper in front of me